java-frontend/src/main/java/org/sonar/java/resolve/TypeAndReferenceSolver.java

```diff
@@ -243,6 +243,28 @@ public void visitMethodInvocation(MethodInvocationTree tree) {
     registerType(tree, returnType);
     if(resolution != null) {
       inferArgumentTypes(argTypes, resolution);
+      List<JavaType> parameterTypes = getParameterTypes(tree.arguments());
+      if(parameterTypes != argTypes) {
+        IdentifierTree identifier;
+        if (methodSelect.is(Tree.Kind.MEMBER_SELECT)) {
+          MemberSelectExpressionTree mset = (MemberSelectExpressionTree) methodSelect;
+          JavaType type = getType(mset.expression());
+          if(type.isTagged(JavaType.DEFERRED)) {
+            throw new IllegalStateException("type of arg should not be defered anymore ??");
+          }
+          identifier = mset.identifier();
+          resolution = resolve.findMethod(methodEnv, type, identifier.name(), parameterTypes, typeParamTypes);
+        } else if (methodSelect.is(Tree.Kind.IDENTIFIER)) {
+          identifier = (IdentifierTree) methodSelect;
+          resolution = resolve.findMethod(methodEnv, identifier.name(), parameterTypes, typeParamTypes);
+        }
+        if(resolution != null && returnType != resolution.type() && resolution.symbol().isMethodSymbol()) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Change this comparison to use the equals method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1698)